node/src/store.rs

```diff
@@ -1,14 +1,326 @@
+use std::fmt::Debug;
+use std::sync::atomic::{AtomicU64, Ordering};
+
+use async_trait::async_trait;
+use celestia_types::ExtendedHeader;
+use dashmap::mapref::entry::Entry;
+use dashmap::DashMap;
+use tendermint::Hash;
+use thiserror::Error;
+use tracing::{info, instrument};
+
+type Result<T, E = StoreError> = std::result::Result<T, E>;
+
+#[async_trait]
+pub trait Store: Send + Sync + Debug {
+    async fn get_by_hash(&self, hash: &Hash) -> Result<ExtendedHeader>;
+    async fn get_by_height(&self, height: u64) -> Result<ExtendedHeader>;
+    async fn get_head(&self) -> Result<ExtendedHeader>;
+
+    async fn height(&self) -> u64;
+    async fn has(&self, hash: &Hash) -> bool;
+    async fn has_at(&self, height: u64) -> bool;
+
+    async fn append_single(&self, header: ExtendedHeader) -> Result<()>;
+
+    async fn append<I>(&self, headers: I) -> Result<()>
+    where
+        I: IntoIterator<Item = ExtendedHeader> + Send,
+        <I as IntoIterator>::IntoIter: Send,
+    {
+        for header in headers.into_iter() {
+            self.append_single(header).await?;
+        }
+
+        Ok(())
+    }
+}
+
 #[derive(Debug)]
-pub struct Store {}
+pub struct InMemoryStore {
+    headers: DashMap<Hash, ExtendedHeader>,
+    height_to_hash: DashMap<u64, Hash>,
+    head_height: AtomicU64,
+}
 
-impl Store {
+#[derive(Error, Debug, PartialEq)]
+pub enum StoreError {
+    #[error("Hash {0} already exists in store")]
+    HashExists(Hash),
+
+    #[error("Height {0} already exists in store")]
+    HeightExists(u64),
+
+    #[error("Failed to append header at height {1}, current head {0}")]
+    NonContinuousAppend(u64, u64),
+
+    #[error("Header not found in store")]
+    NotFound,
+
+    #[error("Store in inconsistent state, lost head")]
+    LostStoreHead,
+
+    #[error("Store in inconsistent state; height->hash mapping exists, {0} missing")]
+    LostHash(Hash),
+}
+
+impl InMemoryStore {
     pub fn new() -> Self {
-        Store {}
+        InMemoryStore {
+            headers: DashMap::new(),
+            height_to_hash: DashMap::new(),
+            head_height: AtomicU64::new(0),
+        }
+    }
+
+    pub fn with_genesis(genesis: ExtendedHeader) -> Self {
+        let genesis_hash = genesis.hash();
+        let genesis_height = genesis.height().value();
+
+        InMemoryStore {
+            headers: DashMap::from_iter([(genesis_hash, genesis)]),
+            height_to_hash: DashMap::from_iter([(genesis_height, genesis_hash)]),
+            head_height: AtomicU64::new(genesis_height),
+        }
+    }
+
+    pub fn get_head_height(&self) -> u64 {
+        self.head_height.load(Ordering::Acquire)
+    }
+
+    #[instrument(err)]
+    pub fn append_continuous(&self, header: ExtendedHeader) -> Result<(), StoreError> {
```

Comment:
Added `_unverified` suffix as well as doc comment with function expectations about responsibilities + TODO so that we can put verification/validation in the right place once syncer crystalises.